### ISSUE TITLE
performance improvements for sub.formatJSON.pl

### ISSUE THREAD

**Issue by trackpete** [trackpete/exiletools-indexer#132]
Analysis for Issue #131 has indicated there are some serious performance bottlenecks in `sub.formatJSON.pl`. This ticket will contain some notes on initial analysis of this subroutine based on consuming from kafka.

As a baseline, here is the initial processing performance without calling this subroutine:

* 46783.4680072929 items/s

And here is the performance of the subroutine as it currently stands when items are processed fully:

* 484.260271632997 items/s

I expect processing to be somewhat heavy, but a factor of ten seems more than expected.

Immediately returning within this subroutine returns the following result:

* 42596.4780497658 items/s

This indicates a surprising amount of overhead in simply calling the subroutine - this bears out some general information on perl, but I didn't expect it to actually be noticeable. I guess when you are calling it 40k times per second it adds up.

Even more interesting, the performance drops drastically the moment we deference the incoming data hash:

* 29354.4608290092 items/s

```
This is currently set as local to share this hash with other subroutines that are called, but even that (local vs my) has a significant overhead. Dereferencing that hash with my instead of local gives us:

* 33684.697395475 items/s
```

If we return on line 20 after pulling the four variables from @_ in the subroutine, we're already dropping again:

* 26214.0440070829 items/s

At this point we've lost 16k items/second (38%) *just by calling a subroutine and populating it with data* for each item. 

The analysis continues...

**Comment by trackpete** [comment]
With a bunch of heavy optimization, I was able to bring full formatting of items (including mod parsing and all that stuff) up to a much more reasonable speed:

* 4081.6750028228 items/s

However, a big cost here is base item identification! Enabling base item name determination with the current method is insane, bringing the speed down to:

* 505.281537034034 items/s

The way this currently works is admittedly clumsy, as I maintain a hash of all known base items and their types, and have to constantly iterate through this hash to do a regexp match against the typeLine mod in order to determine an underlying item. It's super clumsy, and there should be a way to speed this up.

**Comment by thirdy** [comment]
by the way, for ExileTrade, we dont use the baseItemType (sorry I requested
this) but we just use the typeline instead:

https://docs.google.com/spreadsheets/d/1jG2gzYuAukoJtYonlWghbkk9m5W6yGzM21cscpqJ5TU/edit#gid=207179951

On Sun, Apr 10, 2016 at 3:50 AM, Pete Waterman <notifications@github.com>
wrote:

> With a bunch of heavy optimization, I was able to bring full formatting of
> items (including mod parsing and all that stuff) up to a much more
> reasonable speed:
>
>    - 4081.6750028228 items/s
>
> However, a big cost here is base item identification! Enabling base item
> name determination with the current method is insane, bringing the speed
> down to:
>
>    - 505.281537034034 items/s
>
> The way this currently works is admittedly clumsy, as I maintain a hash of
> all known base items and their types, and have to constantly iterate
> through this hash to do a regexp match against the typeLine mod in order to
> determine an underlying item. It's super clumsy, and there should be a way
> to speed this up.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/trackpete/exiletools-indexer/issues/132#issuecomment-207843321>
>



-- 
--
http://sites.google.com/site/thirdyderivera/

**Comment by trackpete** [comment]
The problem with using `typeLine` is that it is not universally accurate thanks to the way GGG randomly populates stuff. For example, small sample output from a quick check:

```
Arcing Sapphire Ring of the Prism : Arcing Sapphire Ring of the Prism is confusing
Beryl Quarterstaff of Anger : Beryl Quarterstaff of Anger is confusing
Bloodthirsty Eternal Sword : Bloodthirsty Eternal Sword is confusing
Blue Imbued Wand : Blue Imbued Wand is confusing
Blurred Diamond Ring of the Goliath : Blurred Diamond Ring of the Goliath is confusing
Burnished Citadel Bow of the Volcano : Burnished Citadel Bow of the Volcano is confusing
Burnished Infernal Sword of Light : Burnished Infernal Sword of Light is confusing
Catalysing Iron Ring of the Pupil : Catalysing Iron Ring of the Pupil is confusing
Dauntless Onyx Amulet of the Infinite : Dauntless Onyx Amulet of the Infinite is confusing
Discharging Amber Amulet of the Augur : Discharging Amber Amulet of the Augur is confusing
Elreon's Amber Amulet : Elreon's Amber Amulet is confusing
Elreon's Amethyst Ring : Elreon's Amethyst Ring is confusing
Elreon's Coral Ring : Elreon's Coral Ring is confusing
Elreon's Gold Amulet : Elreon's Gold Amulet is confusing
Elreon's Gold Ring : Elreon's Gold Ring is confusing
Elreon's Iron Ring of Raiding : Elreon's Iron Ring of Raiding is confusing
Elreon's Iron Ring of Variegation : Elreon's Iron Ring of Variegation is confusing
```

Additionally, almost ALL Jewels are broken (name is only in the typeLine) as are many Flasks. Thus, you have to be very careful using the typeLine.

That said, let's see if I can break down some of these lookup tables into something smaller for these sections. Hrm.

**Comment by trackpete** [comment]
So, first of all, due to a format layout of poedb, I wasn't pulling in all the new base item types. I've updated that.

Then, after cleaning up some bugs and improving performance by forcing base item names for Gems, Cards, Currency, and Quest Items, the only items generally not being identified correctly are:

* Flasks
* Jewels
* Maps (these aren't in item base types at the moment)
* and all the random weird typeLine items

So, with these improvements in-line, here's our new baseline:

* 688.58118204121 items/s

Next, I broke out Flask and Jewel names into separate arrays, as these are much smaller than the full type, and forced Maps to just have a base name of `A Map` to avoid that hassle for now. During the appropriate sections where items are identified as Flasks or Jewels, we only search against those smaller arrays. 

The result:

* 1000.40025127392 items/s

Right now the only problems coming through are with Maeltrom Staffs, I gotta do something about those next.
